core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -112,38 +112,46 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
 
     val principal = session.principal
     val host = session.clientAddress.getHostAddress
-    val acls = getMatchingAcls(resource.resourceType, resource.name)
-
-    // Check if there is any Deny acl match that would disallow this operation.
-    val denyMatch = aclMatch(operation, resource, principal, host, Deny, acls)
-
-    // Check if there are any Allow ACLs which would allow this operation.
-    // Allowing read, write, delete, or alter implies allowing describe.
-    // See #{org.apache.kafka.common.acl.AclOperation} for more details about ACL inheritance.
-    val allowOps = operation match {
-      case Describe => Set[Operation](Describe, Read, Write, Delete, Alter)
-      case DescribeConfigs => Set[Operation](DescribeConfigs, AlterConfigs)
-      case _ => Set[Operation](operation)
-    }
-    val allowMatch = allowOps.exists(operation => aclMatch(operation, resource, principal, host, Allow, acls))
 
-    //we allow an operation if a user is a super user or if no acls are found and user has configured to allow all users
-    //when no acls are found or if no deny acls are found and at least one allow acls matches.
-    val authorized = isSuperUser(operation, resource, principal, host) ||
-      isEmptyAclAndAuthorized(operation, resource, principal, host, acls) ||
-      (!denyMatch && allowMatch)
+    val authorized = if (isSuperUser(operation, resource, principal, host)) {
+      // Check if the user is a superuser to avoid needlessly checking all ACLs
+      // for superusers
+      true
+    } else {
```

Comment:
consider moving the 'else' into a separate method, then this becomes, (I don't think any comments are required on such obvious logic):

```
val authorized = isSuperUser(...) || aclsAllowAccess(...)
```

Following the same pattern, `aclsAllowAccess` could be:

```
val acls = getMatchingAcls(resource.resourceType, resource.name)
val authorized = isEmptyAclAndAuthorized(...) || (!denyAclsExist(acls, ...) && alllowAclsExist(acls, ...))
```

This breaking out into different methods will likely lead to move readable code and less reliance on comments to explain what's happening.  These methods can be local to `authorize`, which I think ends up with something like:

```
override def authorize(session: Session, operation: Operation, resource: Resource): Boolean = {
    if (resource.patternType != PatternType.LITERAL) {
      throw new IllegalArgumentException("Only literal resources are supported. Got: " + resource.patternType)
    }

    val principal = session.principal
    val host = session.clientAddress.getHostAddress

    def isEmptyAclAndAuthorized(acls: Set[Acl]): Boolean = {
      if (acls.isEmpty) {
        authorizerLogger.debug(s"No acl found for resource $resource, authorized = $shouldAllowEveryoneIfNoAclIsFound")
        shouldAllowEveryoneIfNoAclIsFound
      } else false
    }
    
    def denyAclsExist(acls: Set[Acl]) = {
      aclMatch(operation, resource, principal, host, Deny, acls)
    }

    def allowAclsExist(acls: Set[Acl]) = {
      // Check if there are any Allow ACLs which would allow this operation.
      // Allowing read, write, delete, or alter implies allowing describe.
      // See #{org.apache.kafka.common.acl.AclOperation} for more details about ACL inheritance.
      val allowOps = operation match {
        case Describe => Set[Operation](Describe, Read, Write, Delete, Alter)
        case DescribeConfigs => Set[Operation](DescribeConfigs, AlterConfigs)
        case _ => Set[Operation](operation)
      }

      allowOps.exists(operation => aclMatch(operation, resource, principal, host, Allow, acls))
    }    
    
    def aclsAllowAccess = {
      val acls = getMatchingAcls(resource.resourceType, resource.name)
      isEmptyAclAndAuthorized(acls) || (!denyAclsExist(acls) && allowAclsExist(acls))
    }
    
    val authorized = isSuperUser(operation, resource, principal, host) || aclsAllowAccess
    logAuditMessage(principal, authorized, operation, resource, host)
    authorized
  }
```

Though you'll want to check my quick hacking about!

Also, you might want to check a suitable unit test exists that calls authorize for a super user, where deny acls exist. (which should allow access).

Reply:
Your hacking was pretty much spot on, I only took the liberty to very slightly rename the *AclsExist methods.